android/java/org/chromium/chrome/browser/search_engines/settings/BraveSearchEngineAdapter.java

```diff
@@ -76,35 +97,63 @@ static public TemplateUrl getTemplateUrlByShortName(String name) {
         return null;
     }
 
-    public BraveSearchEngineAdapter(Context context, boolean isPrivate) {
-        super(context);
-        mIsPrivate = isPrivate;
+    @Override
+    public void start() {
+        try {
+            runTemplateUrlServiceWithProfile(() -> {
+                super.start();
+                if (!BraveTemplateUrlServiceFactory.getForProfile(mProfile).isLoaded()) {
+                    // updateActiveDSE needs to be delayed for private because service needs to be loaded if no private tab is opened already
+                    needUpdateActiveDSE = true;
+                }
+            });
+        } catch (IllegalStateException e) {
+            // IllegalStateException indicates that search engine is not available anymore. We just
+            // log an error and allow user to choose another search engine instead.
+            Log.e(TAG, e.getMessage());
+        }
+    }
+
+    @Override
+    public void stop() {
+        runTemplateUrlServiceWithProfile(() -> { super.stop(); });
     }
 
+    // OnClickListener:
+
     @Override
     public void onClick(View view) {
-        super.onClick(view);
+        runTemplateUrlServiceWithProfile(() -> { super.onClick(view); });
 
         if (view.getTag() == null) {
             return;
         }
 
         TemplateUrl templateUrl = (TemplateUrl) getItem((int) view.getTag());
-        setDSEPrefs(templateUrl, mIsPrivate);
+        setDSEPrefs(templateUrl, mProfile);
     }
 
+    // TemplateUrlService.LoadListener
+
     @Override
-    public void start() {
-        updateActiveDSE(mIsPrivate);
-        try {
-            super.start();
-        } catch (IllegalStateException e) {
-            // IllegalStateException indicates that search engine is not available anymore. We just
-            // log an error and allow user to choose another search engine instead.
-            Log.e(TAG, e.getMessage());
+    public void onTemplateUrlServiceLoaded() {
+        // It is necessary to ensure user's selection is updated on first entering private setting
+        // but it causes updateActiveDSE() to be called twice (once here and
+        // once from SearchEngineTabModelSelectorObserver)
+        if (needUpdateActiveDSE) {
+            needUpdateActiveDSE = false;
+            updateActiveDSE(mProfile);
         }
+        runTemplateUrlServiceWithProfile(() -> { super.onTemplateUrlServiceLoaded(); });
     }
 
+    @Override
+    public void onTemplateURLServiceChanged() {
+        runTemplateUrlServiceWithProfile(() -> { super.onTemplateURLServiceChanged(); });
+    }
+
+    // private boolean locationEnabled(TemplateUrl templateUrl) is private and unused
```

Comment:
Not sure what this comment indicates

Reply:
There is a method in the original adapter but it is not called anywhere so no need to override it. Maybe I shall just delete the line since it confuses others.